### PR TITLE
Check pod status before hook

### LINKED ISSUE #5137: Velero tries to run preHook and postHook on Completed Pods
I've seen the following error in the logs of my Velero backups:

```
time="2022-07-21T04:22:03Z" level=error msg="Error executing hook" backup=velero/four-hourly-20220721041203 error="unable to upgrade connection: container not found (\"compute\")" hookPhase=pre hookSource=annotation hookType=exec logSource="internal/hook/item_hook_handler.go:214" name=virt-launcher-dns01-qnk9j namespace=powerdns resource=pods
time="2022-07-21T04:22:03Z" level=error msg="Error executing hook" backup=velero/four-hourly-20220721041203 error="unable to upgrade connection: container not found (\"compute\")" hookPhase=post hookSource=annotation hookType=exec logSource="internal/hook/item_hook_handler.go:214" name=virt-launcher-dns01-qnk9j namespace=powerdns resource=pods
```

`virt-launcher-dns01-qnk9j` is a Pod that is Completed. The KubeVirt VM has been live migrated to another Pod. The Pod `virt-launcher-dns01-qnk9j` is a leftover that stays in Completed state.

**What did you expect to happen:**

Velero restic backup should not run on Completed Pods.

**Environment:**

- Velero server version: 1.9.0
- Velero features: `<NOT SET>`
- Kubernetes version: 1.21.12
- Kubernetes installer & version: kubeadm 1.21
- Cloud provider or hardware configuration: bare metal
- OS (e.g. from `/etc/os-release`): Debian 11, Linux `5.10.0-15-amd64 #1 SMP Debian 5.10.120-1 (2022-06-09) x86_64 GNU/Linux`

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

### PR DESCRIPTION
**Pull request by cleverhu**
Thank you for contributing to Velero!

# Please add a summary of your change
Skip exec pod command when pod's phase is succeeded or failed.

# Does your change fix a particular issue?
Fixes #5137

# Please indicate you've done the following:

- [x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [x] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

**Comment by cleverhu** [comment]
This is my first pr and I am following `internal/hook/wait_exec_hook_handler.go` to handle the pod phase. I don't know whether it's proper and whether it needs to consider the special situation of container.

File: pkg/podexec/pod_command_executor.go [hunk 1]
```diff
@@ -123,6 +123,12 @@ func (e *defaultPodCommandExecutor) ExecutePodCommand(log logrus.FieldLogger, it
 			"hookTimeout":   localHook.Timeout,
 		},
 	)
+
+	if pod.Status.Phase == corev1api.PodSucceeded || pod.Status.Phase == corev1api.PodFailed {
```
Comment (blackpiglet) [inline pkg/podexec/pod_command_executor.go:127]: Could you please add a test case in `pod_command_executor_test.go` to cover this case?

**Comment by blackpiglet** [comment]
Please add a changlog according to this document: https://velero.io/docs/v1.5/code-standards/#adding-a-changelog

File: pkg/podexec/pod_command_executor.go (hunk 1 above)
Comment (cleverhu) [reply pkg/podexec/pod_command_executor.go:127]: ok, i will try do it and add the changlog, ths

**Comment by cleverhu** [comment]
@blackpiglet i have add the unit test for the case, please take a look.

**Comment by cleverhu** [comment]
@blackpiglet i have add the changlog and the unit test, do you have any spare time to review? Thank you!

**Comment by blackpiglet** [comment]
@cleverhu 
It's not easy to do a UT for this case. 
There are two ways that I can both accept:
* Revert to inital state of this PR. No update of the UT cases. 
* Put the new logic into a function, and do UT like this case `TestEnsureContainerExists`.

**Comment by cleverhu** [comment]
> @cleverhu It's not easy to do a UT for this case. There are two ways that I can both accept:
> 
> * Revert to inital state of this PR. No update of the UT cases.
> * Put the new logic into a function, and do UT like this case `TestEnsureContainerExists`.

I have move the ut to a new func, is it seems ok? The original way was a little rude, and I didn't think it was appropriate. Thank you for your guidance.

**Review by blackpiglet — APPROVED** [review]

**Comment by cleverhu** [comment]
@reasonerjt PTAL

**Comment by haslersn** [comment]
What's the status?

**Comment by haslersn** [comment]
Not stale, because the issue still needs to be fixed.

File: pkg/podexec/pod_command_executor.go (hunk 1 above)
Comment (blackpiglet) [inline pkg/podexec/pod_command_executor.go:127]: Sorry for the late response, but after going through the code again, I think we may need to change the condition check from current status into something like this:

``` go
if pod.Status.Phase !=  corev1api.PodRunning {
    ...
}
```

**Comment by blackpiglet** [comment]
Furthermore, after this PR is put on hold for a while, this cannot be merged directly.
Could you help to merge main branch to align with the latest change?

**Comment by haslersn** [comment]
Not stale, because the issue still needs to be fixed.

**Comment by haslersn** [comment]
Still not stale, because the issue still needs to be fixed.

**Comment by haslersn** [comment]
@ywk253100 I think your review is missing here

**Review by ywk253100 — APPROVED** [review]

**Comment by haslersn** [comment]
@ywk253100 Did you take a look at #6811 which implements this slightly differently? In particular, it logs an error when trying to run a hook on a Failed Pod, so this doesn't go unnoticed. I think having a Failed Pod is typically something which the user didn't intend, so the fact that we cannot run the hook there is also something which the user didn't intend and should be treated as an error in the backup process.

File: pkg/podexec/pod_command_executor_test.go [hunk 2]
```diff
@@ -262,6 +262,37 @@ func TestEnsureContainerExists(t *testing.T) {
 	assert.NoError(t, err)
 }
 
+func TestPodCompeted(t *testing.T) {
```
Comment (haslersn) [inline pkg/podexec/pod_command_executor_test.go:265]: Typo

**Pull request merged**